File: src/test/java/no/hvl/tk/visual/debugger/util/DiagramToXMLConverterTest.java
package no.hvl.tk.visual.debugger.util;

import no.hvl.tk.visual.debugger.domain.*;
import org.junit.jupiter.api.Test;

import java.util.UUID;

import static org.hamcrest.CoreMatchers.is;
import static org.hamcrest.MatcherAssert.assertThat;
import static org.mockito.Mockito.mockStatic;
import static org.mockito.Mockito.when;

class DiagramToXMLConverterTest {

    @Test
    void emptyDiagram() {
        final ObjectDiagram empty = new ObjectDiagram();

        final String xml = DiagramToXMLConverter.toXml(empty);

        assertThat(xml, is("<?xml version=\"1.0\" encoding=\"UTF-8\" standalone=\"yes\"?>\n<objectDiagram/>\n"));
    }

    @Test
    void primitiveRootValues() {
        final ObjectDiagram diagram = new ObjectDiagram();
        diagram.addPrimitiveRootValue(new ODPrimitiveRootValue("varName1", "varType1", "varValue1"));
        diagram.addPrimitiveRootValue(new ODPrimitiveRootValue("varName2", "varType2", "varValue2"));

        final String xml = DiagramToXMLConverter.toXml(diagram);

        assertThat(xml, is("<?xml version=\"1.0\" encoding=\"UTF-8\" standalone=\"yes\"?>\n" +
                "<objectDiagram>\n" +
                "    <primitiveRootValues>\n" +
                "        <variableName>varName1</variableName>\n" +
                "        <type>varType1</type>\n" +
                "        <value>varValue1</value>\n" +
                "    </primitiveRootValues>\n" +
                "    <primitiveRootValues>\n" +
                "        <variableName>varName2</variableName>\n" +
                "        <type>varType2</type>\n" +
                "        <value>varValue2</value>\n" +
                "    </primitiveRootValues>\n" +
                "</objectDiagram>\n"));
    }

    @Test
    void objectsWithAttributes() {
        final ObjectDiagram diagram = new ObjectDiagram();
        final ODObject obj1 = new ODObject(1, "type", "varName");
        obj1.addAttribute(new ODAttributeValue("attrName", "attrType", "attrValue"));
        diagram.addObject(obj1);

        final String xml = DiagramToXMLConverter.toXml(diagram);

        assertThat(xml, is("<?xml version=\"1.0\" encoding=\"UTF-8\" standalone=\"yes\"?>\n" +
                "<objectDiagram>\n" +
                "    <objects>\n" +
                "        <type>type</type>\n" +
                "        <variableName>varName</variableName>\n" +
                "        <attributeValues>\n" +
                "            <attributeName>attrName</attributeName>\n" +
                "            <attributeType>attrType</attributeType>\n" +
                "            <attributeValue>attrValue</attributeValue>\n" +
                "        </attributeValues>\n" +
                "        <id>1</id>\n" +
                "    </objects>\n" +
                "</objectDiagram>\n"));
    }

    @Test
    void objectsAndLinks() {
        final ObjectDiagram diagram = new ObjectDiagram();
        final ODObject obj1 = new ODObject(1, "type1", "varName1");
        final ODObject obj2 = new ODObject(2, "type2", "varName2");
        diagram.addObject(obj1);
        diagram.addObject(obj2);
        // random UUID's have to be mocked
        //noinspection ResultOfMethodCallIgnored
        mockStatic(UUID.class);
        // Set UUID's to 00000000-0000-0001-0000-000000000001
        when(UUID.randomUUID()).thenReturn(new UUID(1, 1));
        diagram.addLink(new ODLink(obj1, obj2, "friend"));
        diagram.addLink(new ODLink(obj2, obj1, "enemy"));

        final String xml = DiagramToXMLConverter.toXml(diagram);

        assertThat(xml, is("<?xml version=\"1.0\" encoding=\"UTF-8\" standalone=\"yes\"?>\n" +
                "<objectDiagram>\n" +
                "    <objects>\n" +
                "        <type>type1</type>\n" +
                "        <variableName>varName1</variableName>\n" +
                "        <id>1</id>\n" +
                "    </objects>\n" +
                "    <objects>\n" +
                "        <type>type2</type>\n" +
                "        <variableName>varName2</variableName>\n" +
                "        <id>2</id>\n" +
                "    </objects>\n" +
                "    <links>\n" +
                "        <id>00000000-0000-0001-0000-000000000001</id>\n" +
                "        <type>friend</type>\n" +
                "        <from>1</from>\n" +
                "        <to>2</to>\n" +
                "    </links>\n" +
                "    <links>\n" +
                "        <id>00000000-0000-0001-0000-000000000001</id>\n" +
                "        <type>enemy</type>\n" +
                "        <from>2</from>\n" +
                "        <to>1</to>\n" +
                "    </links>\n" +
                "</objectDiagram>\n"));
    }
}

File: src/test/java/no/hvl/tk/visual/debugger/util/DiagramToXMLConverterTest.kt
package no.hvl.tk.visual.debugger.util

import no.hvl.tk.visual.debugger.domain.*
import no.hvl.tk.visual.debugger.util.DiagramToXMLConverter.toXml
import org.hamcrest.CoreMatchers.*
import org.hamcrest.MatcherAssert
import org.junit.jupiter.api.Test

internal class DiagramToXMLConverterTest {
  @Test
  fun emptyDiagram() {
    val empty = ObjectDiagram()

    val xml = toXml(empty)

    MatcherAssert.assertThat(
        xml,
        `is`(
            "<?xml version=\"1.0\" encoding=\"UTF-8\" standalone=\"yes\"?>\n" +
                "<db:objectDiagram xmlns:db=\"http://tk/schema/db\"/>\n"))
  }

  @Test
  fun primitiveRootValues() {
    val diagram = ObjectDiagram()
    diagram.addPrimitiveRootValue(ODPrimitiveRootValue("varName1", "varType1", "varValue1"))
    diagram.addPrimitiveRootValue(ODPrimitiveRootValue("varName2", "varType2", "varValue2"))

    val xml = toXml(diagram)

    MatcherAssert.assertThat(
        xml,
        `is`(
            """
                                   <?xml version="1.0" encoding="UTF-8" standalone="yes"?>
                                   <db:objectDiagram xmlns:db="http://tk/schema/db">
                                       <db:primitiveRootValue variableName="varName1" type="varType1" value="varValue1"/>
                                       <db:primitiveRootValue variableName="varName2" type="varType2" value="varValue2"/>
                                   </db:objectDiagram>
                                   
                                   """
                .trimIndent()))
  }

  @Test
  fun objectsWithAttributes() {
    val diagram = ObjectDiagram()
    val obj1 = ODObject(1, "type", "varName")
    obj1.addAttribute(ODAttributeValue("attrName", "attrType", "attrValue"))
    diagram.addObject(obj1)

    val xml = toXml(diagram)

    MatcherAssert.assertThat(
        xml,
        `is`(
            """
                                   <?xml version="1.0" encoding="UTF-8" standalone="yes"?>
                                   <db:objectDiagram xmlns:db="http://tk/schema/db">
                                       <db:object type="type" variableName="varName" id="Object_1">
                                           <db:attributeValue name="attrName" type="attrType" value="attrValue"/>
                                       </db:object>
                                   </db:objectDiagram>
                                   
                                   """
                .trimIndent()))
  }

  @Test
  fun objectsAndLinks() {
    val diagram = ObjectDiagram()
    val obj1 = ODObject(1, "type1", "varName1")
    val obj2 = ODObject(2, "type2", "varName2")
    diagram.addObject(obj1)
    diagram.addObject(obj2)
    diagram.addLink(ODLink(obj1, obj2, "friend"))
    diagram.addLink(ODLink(obj2, obj1, "enemy"))

    val xml = toXml(diagram)

    MatcherAssert.assertThat(
        xml,
        `is`(
            """
                                   <?xml version="1.0" encoding="UTF-8" standalone="yes"?>
                                   <db:objectDiagram xmlns:db="http://tk/schema/db">
                                       <db:object type="type1" variableName="varName1" id="Object_1"/>
                                       <db:object type="type2" variableName="varName2" id="Object_2"/>
                                       <db:link type="friend" from="Object_1" to="Object_2" id="Link_Object_1_to_Object_2_type_friend"/>
                                       <db:link type="enemy" from="Object_2" to="Object_1" id="Link_Object_2_to_Object_1_type_enemy"/>
                                   </db:objectDiagram>
                                   
                                   """
                .trimIndent()))
  }
}
